packages/tools/src/utilities/segmentation/InterpolationManager/InterpolationManager.ts

```diff
@@ -0,0 +1,173 @@
+  utilities as csUtils,
+  getRenderingEngine,
+} from '@cornerstonejs/core';
+import type { Types } from '@cornerstonejs/core';
+import {
+  AnnotationCompletedEventType,
+  AnnotationModifiedEventType,
+  AnnotationRemovedEventType,
+} from '../../../types/EventTypes';
+import getInterpolationDataCollection from '../../contours/interpolation/getInterpolationDataCollection';
+import type { InterpolationViewportData } from '../../../types/InterpolationTypes';
+import interpolate from '../../contours/interpolation/interpolate';
+import updateRelatedAnnotations from './updateRelatedAnnotations';
+import deleteRelatedAnnotations from './deleteRelatedAnnotations';
+import { InterpolationROIAnnotation } from '../../../types/ToolSpecificAnnotationTypes';
+import ChangeTypes from '../../../enums/ChangeTypes';
+
+const { uuidv4, isEqual } = csUtils;
+
+function getSliceData(viewport): Types.ImageSliceData {
+  const sliceData: Types.ImageSliceData = {
+    numberOfSlices: viewport.getNumberOfSlices(),
+    imageIndex: viewport.getCurrentImageIdIndex(),
+  };
+  return sliceData;
+}
+
+function getMatchingViewport(annotation: InterpolationROIAnnotation) {
+  const { metadata } = annotation;
+  const enabledElement = getEnabledElements().find((enabledElement) => {
+    if (enabledElement.FrameOfReferenceUID === metadata.FrameOfReferenceUID) {
+      const viewport = enabledElement.viewport;
+      const { viewPlaneNormal, viewUp } = viewport.getCamera();
+      return (
+        isEqual(viewPlaneNormal, metadata.viewPlaneNormal) &&
+        isEqual(viewUp, metadata.viewUp)
+      );
+    }
+    return;
+  });
+  return enabledElement?.viewport;
+}
+
+export default class InterpolationManager {
+  static toolNames = [];
+
+  static addTool(toolName: string) {
+    if (-1 === this.toolNames.indexOf(toolName)) {
+      this.toolNames.push(toolName);
+    }
+  }
+
+  static handleAnnotationCompleted = (evt: AnnotationCompletedEventType) => {
+    const { renderingEngineId, viewportId } = evt.detail;
+    const annotation = evt.detail.annotation as InterpolationROIAnnotation;
+    if (!annotation?.metadata) {
+      return;
+    }
+    const { toolName } = annotation.metadata;
+
+    if (-1 === this.toolNames.indexOf(toolName)) {
+      return;
+    }
+
+    const renderingEngine = getRenderingEngine(renderingEngineId);
+    const viewport = renderingEngine.getViewport(viewportId);
+    const sliceData: Types.ImageSliceData = getSliceData(viewport);
+    const viewportData: InterpolationViewportData = {
+      viewport,
+      sliceData,
+      annotation,
+      interpolationUID: annotation.interpolationUID,
+    };
+    const isInitializeLabel = !annotation.interpolationUID;
+    // If any update, triggered on an annotation, then it will be treated as non-autogenerated.
+    annotation.autoGenerated = false;
+    if (!isInitializeLabel) {
+      updateRelatedAnnotations(viewportData, true);
+    }
+    if (!annotation.interpolationUID) {
```

Comment:
I prefer 

if (annotatoin.interpolationUID){
  return
}

// rest of logic

to reduce indentation